pkg/operators/common/systemdcache.go

```diff
@@ -0,0 +1,219 @@
+// Copyright 2023 The Inspektor Gadget authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package common
+
+import (
+	"context"
+	"fmt"
+	"sync"
+	"time"
+
+	"github.com/coreos/go-systemd/v22/dbus"
+	"github.com/sirupsen/logrus"
+
+	"github.com/inspektor-gadget/inspektor-gadget/pkg/container-utils/cgroups"
+)
+
+type SystemdSubInterface interface {
+	AddCgroup(string, uint64)
+	RemoveCgroup(string, uint64)
+}
+
+type SystemdCache struct {
+	dbusConn *dbus.Conn
+
+	idToUnitname sync.Map
+
+	subscribers []SystemdSubInterface
+	subMutex    sync.Mutex
+
+	exit           chan struct{}
+	ticker         *time.Ticker
+	tickerDuration time.Duration
+
+	useCount      int
+	useCountMutex sync.Mutex
+}
+
+var (
+	systemdCache *SystemdCache
+	systemdOnce  sync.Once
+)
+
+func GetSystemdCache() (*SystemdCache, error) {
+	systemdOnce.Do(func() {
+		systemdCache, err = newSystemdCache(1 * time.Second)
+	})
+	return systemdCache, err
+}
+
+func (cache *SystemdCache) Subscribe(sub SystemdSubInterface, publishOldEntries bool) {
```

Comment:
What about making `Subscribe` return the list of the current entries as we do for the [container-collection](https://github.com/inspektor-gadget/inspektor-gadget/blob/4ee22a5d28610cf664896b650936e4f87936d992/pkg/container-collection/container-collection.go#L478)?

Reply:
I like this approach more since we only need to write this code once in the systemdcache (now cgroup-hook). Else we need to handle the list that we get back in every code place where we use the systemdcache/cgroup-hook